### PR TITLE
ref: optimize ProxyRegistry counting

### PR DESCRIPTION
**Pull request by allenlsy**
**Description**:

* Use a counter with mutex to get the size of ProxyRegistry
instead of looping the whole registry
* Use a normal map instead of sync.Map to store proxies

Resolves https://github.com/openservicemesh/osm/issues/4783

**Testing done**:

* Unit test

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| Control Plane              | [X] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

No

2. Is this a breaking change?

No

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)?

No

### CODE REVIEW

File: pkg/envoy/registry/types.go
Comment (keithmattix) [inline pkg/envoy/registry/types.go:18]: A raw map is not threadsafe. A consequence of this change is that the locks protecting this data structure are held longer. How confident are we that this change results in higher performance?

**Review by keithmattix — CHANGES_REQUESTED** [review]
The rationale for this change isn't clear to me. Is looping over the proxyregistry causing some bottleneck? What other approaches might be taken to achieve a similar performance result?

File: pkg/envoy/registry/types.go
Comment (allenlsy) [reply pkg/envoy/registry/types.go:18]: I have similar doubt actually. I suggest doing a benchmark to compare these two approaches.

Comment (steeling) [reply pkg/envoy/registry/types.go:18]: there's no need to hold the lock longer, but the locks are locking on a single int, instead of on the (non-threadsafe) map itself.

It would be faster (no benchmark needed), simply grab the lock whenever reading/writing to the map

File: pkg/envoy/registry/registry.go
Comment (steeling) [inline pkg/envoy/registry/registry.go:35]: these accesses are not thread safe and will result in a panic

File: pkg/envoy/registry/registry.go
Comment (steeling) [inline pkg/envoy/registry/registry.go:48]: calling this while the lock is held will result in a deadlock

File: pkg/envoy/registry/types.go
Comment (steeling) [inline pkg/envoy/registry/types.go:22]: standard practice is to call it `mu sync.Mutex`

and simply place the protected fields underneath it, in this case the map itself. No need to store the length, you can just len whenever you need it. that's a fast op

File: pkg/envoy/registry/types.go
Comment (keithmattix) [reply pkg/envoy/registry/types.go:18]: My comment about holding the lock longer was referring to the PR's implementation since the lock was being acquired and then not released until the count was updated.

Are we sure this is a hot path though? How often are we checking the length of this map?

Comment (allenlsy) [reply pkg/envoy/registry/types.go:18]: I did a test comparing this the normal map v.s. original sync.Map. The test consists of two parts:

* Add 10000 proxies to the registry with 10000 goroutines
* When registry contains 10000 proxies, get the proxies count.

Both tests ensured the correctness of proxies count.

|  | sync.Map | normal map with mutex |
| --- | --- | --- |
| Add 10000 proxies | 9.5s | 0.26s |
| GetCount | 957642 ns | 13ns  |

Comment (keithmattix) [reply pkg/envoy/registry/types.go:18]: Great, thanks @allenlsy. That's what I was looking for

File: pkg/envoy/registry/types.go
Comment (keithmattix) [reply pkg/envoy/registry/types.go:22]: +1

File: pkg/envoy/registry/registry.go
Comment (steeling) [inline pkg/envoy/registry/registry.go:60]: there's actually 2 locks associated with a read lock. Read locks are great for when you do a lot of reads, and your ops (read/write), might have some longer blocking function. In this case, since it's just setting/deleting/reading values it should be faster to use a standard lock

File: pkg/envoy/registry/types.go
Comment (steeling) [inline pkg/envoy/registry/types.go:24]: see not above about standard mutex vs rw

**Comment by steeling** [comment]
hey @allenlsy friendly reminder not to squash commits to make reviews easier

File: pkg/envoy/registry/announcement_handlers.go
Comment (steeling) [inline pkg/envoy/registry/announcement_handlers.go:38]: leverage GetCOnnected proxy here, since this needs to grab the lock

File: pkg/envoy/registry/registry_benchmark.go
Comment (steeling) [inline pkg/envoy/registry/registry_benchmark.go:1]: benchmark tests still belong int `*_test.go` files, so as not to be included in non test files

File: pkg/envoy/registry/registry_test.go
Comment (steeling) [inline pkg/envoy/registry/registry_test.go:47]: Not going to block on this, but would be nice to move away from ginko in unit tests

File: pkg/envoy/registry/types.go
Comment (steeling) [inline pkg/envoy/registry/types.go:18]: nit: place the connected proxies field under the mutex field, which signifies what the mutex locks.

File: pkg/envoy/registry/registry_test.go
Comment (keithmattix) [reply pkg/envoy/registry/registry_test.go:47]: +1 table driven tests or at least testify assert is more idiomatic

Comment (trstringer) [reply pkg/envoy/registry/registry_test.go:47]: Also a +1 on this! We are moving away from ginkgo for testing, so we shouldn't be introducing more at this point.

File: pkg/envoy/registry/registry.go
Comment (allenlsy) [reply pkg/envoy/registry/registry.go:52]: In our test, after removing lock, concurrent map read and write is found. I think the program may add and delete the same key simultaneously.

**Comment by allenlsy** [comment]
Ping here for the review @steeling @keithmattix

File: pkg/envoy/ads/response_benchmark_test.go
Comment (steeling) [inline pkg/envoy/ads/response_benchmark_test.go:190]: you intend to remove this?

File: pkg/envoy/registry/registry.go
Comment (steeling) [inline pkg/envoy/registry/registry.go:20]: The way in which a proxy can disconnect and reconnect in rapid succession would actually allow a proxy to still be "registered", prior to it being deregistered.

This is a pre-existing bug, and isn't quite in the critical path, so likely not a big concern, although we should make a note of it and consider addressing it later.

One solution would require a nonce, or a version number to check when unregistering, and only unregistering if those match.

File: pkg/envoy/registry/registry.go
Comment (steeling) [inline pkg/envoy/registry/registry.go:37]: nit: no need for nil check, a bit simpler:

```go
pr.mu.Lock()
defer pr.mu.Unlock()
return pr.connectedProxies[uuid]
```

File: pkg/envoy/registry/registry.go
Comment (steeling) [inline pkg/envoy/registry/registry.go:46]: nit: this log doesn't add much value IMO, can just simply call delete within the lock

File: pkg/envoy/ads/response_benchmark_test.go
Comment (allenlsy) [reply pkg/envoy/ads/response_benchmark_test.go:190]: Yes. It's not necessary in this case since it's the end of the test.

File: pkg/envoy/registry/registry.go
Comment (allenlsy) [reply pkg/envoy/registry/registry.go:20]: Let's create a new issue for this bug. Nonce might not be a good solution. Sequential version number should maintain the request sequence. 

E.g. a proxy sends unregister and register requests. Due to network delay, register request arrives earlier. Without sequence number, it will eventually be deregistered.

**Review by keithmattix — APPROVED** [review]

File: pkg/envoy/registry/registry.go
Comment (jaellio) [reply pkg/envoy/registry/registry.go:20]: Can you create the issue and add a TODO here referencing the issue number?

**Review by jaellio — APPROVED** [review]
@allenlsy What is the force rerun file included in this PR? Waiting on @steeling's review to merge.

**Comment by allenlsy** [comment]
@jaellio That's an empty file commit to force re-run a flaky e2e test. I can squash all commits here into a single one pre-merge

**Comment by jaellio** [comment]
> @jaellio That's an empty file commit to force re-run a flaky e2e test. I can squash all commits here into a single one pre-merge

Thanks for clarifying! Is this something we want to add in this PR or in a separate PR? Should we review the flaky test?

**Comment by allenlsy** [comment]
> > @jaellio That's an empty file commit to force re-run a flaky e2e test. I can squash all commits here into a single one pre-merge
> 
> Thanks for clarifying! Is this something we want to add in this PR or in a separate PR? Should we review the flaky test?

Based on the failed test [here](https://github.com/openservicemesh/osm/runs/7526003652?check_suite_focus=true#step:5:263), I don't think it's related to the changes in this PR. We should address it separately.

**Review by steeling — APPROVED** [review]
approved after comments are addressed, thanks Allen!

File: pkg/envoy/registry/registry.go
Comment (allenlsy) [reply pkg/envoy/registry/registry.go:20]: New issue opened to track the bug: https://github.com/openservicemesh/osm/issues/4950

Comment (jaellio) [reply pkg/envoy/registry/registry.go:20]: Awesome! Did you add the TODO(#4950) comment?

Comment (allenlsy) [reply pkg/envoy/registry/registry.go:21]: Just added. Let's wait for the CI pass.

**Pull request merged**
